Format Swift sources in `utils` and add a test to enforce their format going forward

AnthonyLatsis opened the pull request:
About time we did this.

AnthonyLatsis committed:
Move `.swift-format` to project root and sync it with swift-syntax

AnthonyLatsis committed:
Add test to enforce format for Swift sources in `utils`

AnthonyLatsis committed:
Format Swift sources in `utils`

AnthonyLatsis commented:
Okay, let’s test this out just on `utils/` for now. None of the Swift sources in there were modified since we branched.

AnthonyLatsis commented:
@swift-ci please smoke test

AnthonyLatsis commented:
@swift-ci please smoke test

eeckstein commented:
I'm not a big fan of this, because such massive changes call for troubles without a big benefit (e.g. git blame will be less informative, merging results in more conflicts). But, if people really want this, I'm fine with it.
However, before we do this with SwiftCompilerSources, we need to fix a few things in swift-format. IIRC, at least two things are not ideal today (I discussed this with @ahoppen some time ago):
1. swift-format converts `private extension T { func foo() }` -> `extension T { fileprivate func foo() }`. This is confusing because I want the whole extension to be private without needing to take care of each members of the extension
2. it generates wrong indentation for If/guard with multi-line conditions, e.g.
```
   guard cond1,
    cond2
  else {
```
All the condition lines should be nicely aligned.

AnthonyLatsis commented:
> > it generates wrong indentation for If/guard with multi-line conditions
> 
> Do you know if we’re tracking this anywhere? That strikes me more like a stylistic alternative than a bug, but either way I do not believe it a good enough reason to pull the pin on this.

We should be able to mark these statements with `// swift-format-ignore` until the issue is resolved, though.

eeckstein commented:
I could easily reproduce it with an older swift-syntax and with top-of-main:
```
private extension Int {
  func foo() {}
}
```
->
```
extension Int {
  fileprivate func foo() {}
}
```

> That strikes me more like a stylistic alternative than a bug
Well, the whole point of swift-format is to improve the formatting. Therefore I consider making it worse a bug. :-)
And it's not an "alternative", because it clearly makes the code less readable.

AnthonyLatsis commented:
> Do you know if we’re tracking this anywhere?

Found it: https://github.com/swiftlang/swift-format/issues/325

eeckstein commented:
> Do you know if we’re tracking this anywhere?

I just created https://github.com/swiftlang/swift-format/issues/1036 and https://github.com/swiftlang/swift-format/issues/1037

> but either way I do not believe it a good enough reason to pull the pin on this.

I don't agree. We should not re-format existing sources if some code patterns look worse than before.
I don't believe it will be a big effort to fix those problems in swift-syntax.

AnthonyLatsis committed:
.swift-format: Enable `DoNotUseSemicolons` & `ReturnVoidInsteadOfEmptyTuple`

AnthonyLatsis committed:
.swift-format: Enable `spacesAroundRangeFormationOperators`

This seems like a better choice in general, e.g. `1 + 2...10` or
(from `utils/gen-unicode-data`) `for j in (idx / 64) & ~7..<idx / 64 {`.

AnthonyLatsis commented:
> I don't believe it will be a big effort to fix those problems in swift-format.

@eeckstein Discussed https://github.com/swiftlang/swift-format/issues/325 with Ben Barham offline. It appears to be a tricky problem. I’m happy to give it a go, but it will likely take a few years before our CI environments are bumped to a 6.3 toolchain (assuming the 6.3-aligned formatter will contain a fix). macOS nodes are currently running 6.0. Would you still prefer to postpone this over turning a blind eye on the formatting of multiline statement conditions?

eeckstein commented:
> I’m happy to give it a go

That would be great!

>  It appears to be a tricky problem

If this is really the case, as a workaround can we post-process the swift-format output with a simple e.g. awk script which fixes the condition indentation?

> Would you still prefer to postpone this over turning a blind eye on the formatting of multiline statement conditions?

For the one-time reformatting we can use whatever swift-format we want. And for continuous verification we should also be able to use the just built swift-format, no?

> I’ve gone ahead and swift-format-ignored the statements where it reindented multiline conditions

This is horrible, IMO.

finagolfin commented:
The vast majority of this pull appears to be fixing white space. I think these Swift files in `utils/` are sufficiently rarely updated that it is worth applying this change and then enforcing the formatting from now on, just my two cents.
